p2p/src/main/java/bisq/network/p2p/peers/BroadcastHandler.java

```diff
@@ -239,36 +185,145 @@ public void onAwakeFromStandby() {
+        if (broadcastRequests.size() == 1) {
+            // If we only have 1 message we avoid the overhead of the BundleOfEnvelopes and send the message directly
+            return broadcastRequests.get(0).getMessage();
+        } else {
+            return new BundleOfEnvelopes(broadcastRequests.stream()
+                    .map(Broadcaster.BroadcastRequest::getMessage)
+                    .collect(Collectors.toList()));
+        }
+    }
 
-        if (listener != null && (numOfCompletedBroadcasts + numOfFailedBroadcasts == numPeers || stopped))
-            listener.onBroadcastCompleted(message, numOfCompletedBroadcasts, numOfFailedBroadcasts);
+    private void maybeNotifyListeners(List<Broadcaster.BroadcastRequest> broadcastRequests) {
+        int numOfCompletedBroadcastsTarget = Math.max(1, Math.min(numPeersForBroadcast, 3));
+        // We use equal checks to avoid duplicated listener calls as it would be the case with >= checks.
+        if (numOfCompletedBroadcasts == numOfCompletedBroadcastsTarget) {
+            // We have heard back from 3 peers (or all peers if numPeers is lower) so we consider the message was sufficiently broadcast.
+            broadcastRequests.stream()
+                    .filter(broadcastRequest -> broadcastRequest.getListener() != null)
+                    .map(Broadcaster.BroadcastRequest::getListener)
+                    .forEach(listener -> listener.onSufficientlyBroadcast(broadcastRequests));
+        } else {
+            // Number of open requests to peers is less than we need to reach numOfCompletedBroadcastsTarget.
+            // Thus we never can reach required resilience as too many numOfFailedBroadcasts occurred.
+            int openRequests = numPeersForBroadcast - numOfCompletedBroadcasts - numOfFailedBroadcasts;
+            int maxPossibleSuccessCases = openRequests + numOfCompletedBroadcasts;
+            // We subtract 1 as we want to have it called only once, with a < comparision we would trigger repeatedly.
+            if (maxPossibleSuccessCases == numOfCompletedBroadcastsTarget - 1) {
+                broadcastRequests.stream()
+                        .filter(broadcastRequest -> broadcastRequest.getListener() != null)
+                        .map(Broadcaster.BroadcastRequest::getListener)
+                        .forEach(listener -> listener.onNotSufficientlyBroadcast(numOfCompletedBroadcasts, numOfFailedBroadcasts));
+            } else if (timeoutTriggered && numOfCompletedBroadcasts < numOfCompletedBroadcastsTarget) {
+                // We did not reach resilience level and timeout prevents to reach it later
+                broadcastRequests.stream()
+                        .filter(broadcastRequest -> broadcastRequest.getListener() != null)
+                        .map(Broadcaster.BroadcastRequest::getListener)
+                        .forEach(listener -> listener.onNotSufficientlyBroadcast(numOfCompletedBroadcasts, numOfFailedBroadcasts));
+            }
```

Comment:
These two cases could be combined to one block

Reply:
Will do...